include/dnscpp/tcp.h

```diff
@@ -103,6 +106,18 @@ class Tcp : public Socket, private Monitor, private Connecting
      */
     bool _connected = false;
 
+    /**
+     *  Query IDs currently sent over the wire
+     *  @var std::set
+     */
+    std::set<uint16_t> _queryids;
+
+    /**
+     *  Queries awaiting to be sent over wire, but had an ID collision with one that is already in flight
+     *  @var std::map
+     */
+    std::map<uint16_t, std::list<std::reference_wrapper<const Query>>> _awaiting;
```

Comment:
Is the reference-wrapper an implicit requirement that the Query stays in scope during the call? I don't think that this is the case (for example when userspace calls the cancel() method). Better make a full copy of the query. Premature optimizations are not needed here.

Reply:
fixed in 5d08f4e